File: Graph/GFG/Medium/Prim'sAlgorithm/Prims.java
// https://www.geeksforgeeks.org/problems/minimum-spanning-tree/1

import java.util.PriorityQueue;

class Pair {

    int wt, node;

    Pair(int wt, int node) {
        this.wt = wt;
        this.node = node;
    }
}
public class Prims {
    static int spanningTree(int V, int E, List<List<int[]>> adj) {
        // Code Here.
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> {
            return a.wt - b.wt;
        });
        boolean visited[] = new boolean[V];

        pq.offer(new Pair(0, 0)); // wt, node, parent
        int sum = 0;
        while (!pq.isEmpty()) {
            Pair curr = pq.poll();
            int node = curr.node;
            if (visited[node]) {
                continue;
            }
            int wt = curr.wt;
            visited[node] = true;
            sum += wt;
            for (int[] neighbor : adj.get(node)) {
                int adjNode = neighbor[0], adjWt = neighbor[1];
                if (!visited[adjNode]) {
                    pq.offer(new Pair(adjWt, adjNode));
                }
            }
        }
        return sum;
    }
}

File: Graph/GFG/Medium/Prim'sAlgorithm/Prims.cpp
// https://www.geeksforgeeks.org/problems/minimum-spanning-tree/1
#include<bits/stdc++.h>
using namespace std;
int spanningTree(int V, vector<vector<int>> adj[])
{
    // code here
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    vector<bool> visited(V, false);

    pq.push({0, 0});
    int sum = 0;
    while (!pq.empty())
    {
        auto curr = pq.top();
        pq.pop();
        int node = curr.second;
        if (visited[node])
            continue;
        int wt = curr.first;
        visited[node] = true;
        sum += wt;
        for (vector<int> &neighbor : adj[node])
        {
            int adjNode = neighbor[0], adjWt = neighbor[1];
            if (!visited[adjNode])
            {
                pq.push({adjWt, adjNode});
            }
        }
    }
    return sum;
}
int main(void) {
    
    return 0;
}